### ISSUE TITLE
Multiplying images of type UnsignedInt16 results in bad overflows

### ISSUE THREAD

**Issue by nicost** [nicost/cl-kernel-bindings#3]
Current code in KernelsTest demonstrates the issue:

>  if (srcImages[i].getNativeType() == NativeTypeEnum.UnsignedShort) {
>        Kernels.set(gCLKE, srcImages[i], 25000.0f);                     
>        Kernels.multiplyImages(gCLKE, srcImages[i], src2, dstImages[i]);
>        minMax = Kernels.minMax(gCLKE, dstImages[i], 36); 
>        Assert.assertEquals(65535.0f, minMax[0], 0.000001);
>  }

Results in value 29.0 rather than the expected 65535.0.

**Comment by haesleinhuepf** [comment]
Hey @nicost ,

thanks for pointing me to this. Feel free to tag me next time if you think I could be of help.

I wrote a [minimal example script](https://github.com/clij/clij-advanced-filters/commit/a9b6a1347cb05a0928ecaac64dcf250171d88483) to confirm the issue in clij. It outputs:

```
Started multiply_images_test.py at Fri Nov 15 08:10:41 CET 2019
Testing GeForce GTX 1070 with Max-Q Design OpenCL Version 1.2 with images of type Float
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 80000.0
Testing GeForce GTX 1070 with Max-Q Design OpenCL Version 1.2 with images of type UnsignedShort
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 29.0
Testing GeForce GTX 1070 with Max-Q Design OpenCL Version 1.2 with images of type UnsignedByte
100 * 2 = 200.0
200 * 2 = 255.0
20000 * 2 = 255.0
40000 * 2 = 255.0
Testing Intel(R) UHD Graphics 630 OpenCL Version 2.0 with images of type Float
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 80000.0
Testing Intel(R) UHD Graphics 630 OpenCL Version 2.0 with images of type UnsignedShort
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 29.0
Testing Intel(R) UHD Graphics 630 OpenCL Version 2.0 with images of type UnsignedByte
100 * 2 = 255.0
200 * 2 = 0.0
20000 * 2 = 0.0
40000 * 2 = 0.0
Testing Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz OpenCL Version 2.0 with images of type Float
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 80000.0
Testing Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz OpenCL Version 2.0 with images of type UnsignedShort
100 * 2 = 200.0
200 * 2 = 400.0
20000 * 2 = 40000.0
40000 * 2 = 29.0
Testing Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz OpenCL Version 2.0 with images of type UnsignedByte
100 * 2 = 200.0
200 * 2 = 255.0
20000 * 2 = 255.0
40000 * 2 = 255.0
bye
```

Thus, I can confirm that NVidia and Intel GPUs (and CPUs) are affected by this bug. Apparently, also ubyte images show similar overflow issues depending on the used OpenCL-device. I will fix it asap in clij-core and let you know.

Cheers,
Robert

**Comment by haesleinhuepf** [comment]
Hi again @nicost,

the bug was in the [preamble.cl](https://github.com/clij/clij-clearcl/commit/7aba614e171ecd2731ef866f335b15b91a2ff6fe) of clij-clearcl. 

Thanks for finding this one! Your help is very much appreciated!

Cheers,
Robert

**Comment by haesleinhuepf** [comment]
FYI, this bugfix is also related: https://github.com/clij/clij-core/commit/d9229ec070ba5f4a1e8ccbdb3feed10ffe27aeb2

**Comment by nicost** [comment]
Fixed by upgrading to clij-clearcl 0.10.2 (commit https://github.com/nicost/cl-kernel-bindings/commit/8d3c3e77f952fb82cead581abd48b4dc6bbaedbc)
